src/resolve/resolved-dns-server.c

```diff
@@ -576,6 +576,29 @@ void dns_server_warn_downgrade(DnsServer *server) {
         server->warned_downgrade = true;
 }
 
+bool dns_server_limited_domains(DnsServer *server)
+{
+        DnsSearchDomain *domain;
+        bool domain_restricted = false;
+
+        /* Check if the server has route-only domains without ~., i. e. whether
+         * it should only be used for particular domains */
+        if (!server->link)
+                return false;
+
+        LIST_FOREACH(domains, domain, server->link->search_domains) {
+                if (domain->route_only) {
+                        domain_restricted = true;
+                        /* ~. means "any domain", thus it is a global server */
+                        if (streq(DNS_SEARCH_DOMAIN_NAME(domain), ".")) {
+                                return false;
+                        }
+                }
+        }
+
+        return domain_restricted;
```

Comment:
Hello Zbigniew, thanks for reviewing!

Zbigniew Jędrzejewski-Szmek [2016-06-29  8:31 -0700]:
> After your change, '~.' doesn't make much sense either.

Good point. Indeed the concept of a "routing only domain list" only makes sense with a set of domains which is neither empty nor everything, as the former is equivalent to "no DNS server" and the latter is just a normal global DNS server. So maybe `~.` should be deprecated altogether? Indeed I was misled twice by this already; the manpage has some explanation about it, but I'm afraid I don't understand it or what its purpose should be. I think we should let @poettering weigh in on that one.

> I think we should step back and first talk how the whole scheme should work, and only then start looking at the nitty gritty details.

Yes, agreed. I mostly put this up to have something to experiment with.  Issue #3420 has some initial discussion, but no conclusion so far.

> My first observation is that normal Domains= settings conflates two things (search suffixes and search domains), while route-only domains only does the second of those. But I would expect completely different matching rules for the two different functionalities of normal Domains settings. For search suffixing, we want to try any and all links, because any of them could contain our single-label name.

Indeed here it gets a bit fuzzy, but I understand that differently. If I have `Domains=lab` on a `tun0` link, then my intuition would be that a query for `foo` would try to resolve `foo` and `foo.lab` on the DNS server(s) of `tun0`, but not try to expand any query with `*.lab` on other interfaces' servers. I. e. a link specific `Domain=` is limited to that link for searching as well.  In none of the use cases that I have at hand on a daily basis (getting search domains from my router over wl*, having a "lxc" search domain on the "lxcbr0" bridge, and getting a few search domains from Canonical over the VPN (tun0)) the search domains are meant to be global. I'm not sure whether there are practical use cases where domain search lists picked up over some link would also apply to other links, but if so, there is no current syntax for that -- maybe *that* is the place which needs a special symbol (but not `~.` please, as that has the wrong connotation).

> For search routing, we want to try the only those that match, usually one, because of the privacy and efficiency issues you mentioned. I think that the confusion about how Domains= routing should work stems from this conflation.

I suppose you mean "For routing only domains, we want..". Indeed that's my specific complaint in these two issues, I didn't change the behaviour of search domains (and if we want to change it this should be a separate PR).

> I think routing should work like this:
> 1. Every link with does not have Domains= specified has an implicit default of `Domains=~.`.

Agreed, i. e. those are "global" nameservers.

> 2. When Domains is used with non-routing-only names, treat the non-tilde suffixes in the following way:
>     - implicitly add a routing-only domain for the specified suffix to this network

Agreed.

>     - if none of the specified suffixes are routing-only, add the implicit ~. as in 1.

I don't think we should do that by default, see above.

> 2. For a query, look over all networks and find the longest matching suffix. Because of 1., we expect to get al least one match[*]. Query on all matched networks and only those.
> 
> [*] It is possible that we only have networks with `Domains=~something`. Those are not general purpose servers, and should only match queries that end in `.something`. In that case it is possible that we get no match and the query fails.

Agreed as well.

> Maybe things are already like that, but I don't think so, neither before your PR nor after.

Handling of routing-only domains should be fairly close now. I haven't checked/changed handling of search domains.